python/pyspark/sql/types.py

```
@@ -57,7 +57,25 @@ def __ne__(self, other):
 
     @classmethod
     def typeName(cls):
-        return cls.__name__[:-4].lower()
+        typeTypeNameMap = {"DataType": "data",
+                           "NullType": "null",
+                           "StringType": "string",
+                           "BinaryType": "binary",
+                           "BooleanType": "boolean",
+                           "DateType": "date",
+                           "TimestampType": "timestamp",
+                           "DecimalType": "decimal",
+                           "DoubleType": "double",
+                           "FloatType": "float",
+                           "ByteType": "byte",
+                           "IntegerType": "integer",
+                           "LongType": "long",
+                           "ShortType": "short",
+                           "ArrayType": "array",
+                           "MapType": "map",
+                           "StructField": "struct",
```

Btw, I don't think `i.typeName()` is a valid usage. We better let it throw an exception when calling `typeName` on `StructField`.

`i.dataType.typeName()` is more reasonable call to me.